pbsmetrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,242 @@
+)
+
+// Defines the actual Prometheus metrics we will be using. Satisfies interface MetricsEngine
+type Metrics struct {
+	Registry      *prometheus.Registry
+	connCounter   prometheus.Gauge
+	connError     *prometheus.CounterVec
+	imps          *prometheus.CounterVec
+	requests      *prometheus.CounterVec
+	reqTimer      *prometheus.HistogramVec
+	adaptRequests *prometheus.CounterVec
+	adaptTimer    *prometheus.HistogramVec
+	adaptBids     *prometheus.CounterVec
+	adaptPrices   *prometheus.HistogramVec
+	cookieSync    prometheus.Counter
+	userID        *prometheus.CounterVec
+}
+
+// NewMetrics constructs the appropriate options for the Prometheus metrics. Needs to be fed the promethus config
+// Its own function to keep the metric creation function cleaner.
+func NewMetrics(cfg config.PrometheusMetrics) *Metrics {
+	// define the buckets for timers
+	timerBuckets := prometheus.LinearBuckets(0.05, 0.05, 20)
+	timerBuckets = append(timerBuckets, []float64{1.5, 2.0, 3.0, 5.0, 10.0, 50.0}...)
+
+	standardLabelNames := []string{"source", "type", "pubid", "browser", "cookie", "status"}
+
+	adapterLabelNames := []string{"source", "type", "pubid", "browser", "cookie", "status", "adapter"}
+	bidLabelNames := []string{"source", "type", "pubid", "browser", "cookie", "status", "adapter", "bidtype", "hasadm"}
+
+	metrics := Metrics{}
+	metrics.Registry = prometheus.NewRegistry()
+	metrics.connCounter = newConnCounter(cfg)
+	metrics.Registry.MustRegister(metrics.connCounter)
+	metrics.connError = newCounter(cfg, "active_connections_total",
+		"Errors reported on the connections coming in.",
+		[]string{"ErrorType"},
+	)
+	metrics.Registry.MustRegister(metrics.connError)
+	metrics.imps = newCounter(cfg, "imps_requested_total",
+		"Total number of impressions requested through PBS.",
+		standardLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.imps)
+	metrics.requests = newCounter(cfg, "requests_total",
+		"Total number of requests made to PBS.",
+		standardLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.requests)
+	metrics.reqTimer = newHistogram(cfg, "request_time_seconds",
+		"Seconds to resolve each PBS request.",
+		standardLabelNames, timerBuckets,
+	)
+	metrics.Registry.MustRegister(metrics.reqTimer)
+	metrics.adaptRequests = newCounter(cfg, "adapter_requests_total",
+		"Number of requests sent out to each bidder.",
+		adapterLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.adaptRequests)
+	metrics.adaptTimer = newHistogram(cfg, "adapter_time_seconds",
+		"Seconds to resolve each request to a bidder.",
+		adapterLabelNames, timerBuckets,
+	)
+	metrics.Registry.MustRegister(metrics.adaptTimer)
+	metrics.adaptBids = newCounter(cfg, "adapter_bids_recieved_total",
+		"Number of bids recieved from each bidder.",
+		bidLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.adaptBids)
+	metrics.adaptPrices = newHistogram(cfg, "adapter_prices",
+		"Value of the highest bids from each bidder.",
```

Comment:
What makes this just "the highest" bids?

Reply:
A quick reading of the code brought me to this conclusion. I think I missed the loop, will need to fix the description, Thanks.